fix(transformer/decorator): do not lose `WeakMap` when decorator and class properties transforms combined

overlookmotel opened the pull request:
Closes #9171.

Prevent injected statements attached to the `Class` statement getting lost when `class C {}` is replaced with `let C = class C {};` by moving them to be attached to the new `let C` statement.

---

I'm actually unclear why we have to transform `class C {}` to `let C = class C {};` anyway. `class C {}` already creates a binding `C` which behaves like a `let` binding, so it can be reassigned.

Input:

```js
@dec
export class C {}
```

Current output:

```js
let C = class C {};
C = babelHelpers.decorate([dec], C);
export { C };
```

Could just be:

```js
export class C {}
C = babelHelpers.decorate([dec], C);
```

[ESBuild does this](https://esbuild.github.io/try/#dAAwLjI0LjAALS10YXJnZXQ9ZXMyMDIwAEBkZWMoKQpjbGFzcyBGb28gewogICNiYXIgPSAiYmFyIjsKICB0ZXN0KCkgeyByZXR1cm4gdGhpcy4jYmFyOyB9Cn0K) - it keeps the class declaration.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9952?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9952** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9952?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9952?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9951** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9951?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9950** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9950?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9952).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing commented:
Your esbuild configuration is wrong, the decorator is treated as ES Decorator, you should set `loader=ts` and turn on `experimentDecorator`

The correct example is
https://esbuild.github.io/try/#dAAwLjI0LjAALS10YXJnZXQ9ZXMyMDIwIC0tbG9hZGVyPXRzICItLXRzY29uZmlnLXJhdz17XCJjb21waWxlck9wdGlvbnNcIjp7XCJleHBlcmltZW50YWxEZWNvcmF0b3JzXCI6dHJ1ZX19IgBAZGVjKCkKY2xhc3MgRm9vIHsKICAjYmFyID0gImJhciI7CiAgdGVzdCgpIHsgcmV0dXJuIHRoaXMuI2JhcjsgfQp9Cg

overlookmotel commented:
> Your esbuild configuration is wrong

Shame on me!

Thanks for pointing that out. Odd that ESBuild produces less optimal output for legacy decorators. But, regardless of what ESBuild does, I think retaining the `class C` declaration would be perfectly valid (and is shorter).

Dunqing commented:
### Merge activity

* **Mar 21, 10:07 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
fix(transformer/decorator): do not lose `WeakMap` when decorator and class properties transforms combined (#9952)

Closes #9171.

Prevent injected statements attached to the `Class` statement getting lost when `class C {}` is replaced with `let C = class C {};` by moving them to be attached to the new `let C` statement.

---

I'm actually unclear why we have to transform `class C {}` to `let C = class C {};` anyway. `class C {}` already creates a binding `C` which behaves like a `let` binding, so it can be reassigned.

Input:

```js
@dec
export class C {}
```

Current output:

```js
let C = class C {};
C = babelHelpers.decorate([dec], C);
export { C };
```

Could just be:

```js
export class C {}
C = babelHelpers.decorate([dec], C);
```

[ESBuild does this](https://esbuild.github.io/try/#dAAwLjI0LjAALS10YXJnZXQ9ZXMyMDIwAEBkZWMoKQpjbGFzcyBGb28gewogICNiYXIgPSAiYmFyIjsKICB0ZXN0KCkgeyByZXR1cm4gdGhpcy4jYmFyOyB9Cn0K) - it keeps the class declaration.
